vnext/Desktop/WinHTTPWebSocketResource.cpp

```diff
@@ -0,0 +1,447 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+// clang-format off
+#include "pch.h"
+
+#include <Unicode.h>
+#include <Utils.h>
+#include "WinHTTPWebSocketResource.h"
+
+// Windows API
+#include <comutil.h>
+
+using Microsoft::Common::Unicode::Utf8ToUtf16;
+
+using std::function;
+using std::size_t;
+using std::string;
+using std::wstring;
+
+namespace Microsoft::React
+{
+
+WinHTTPWebSocketResource::WinHTTPWebSocketResource(const string& urlString)
+  : m_openSessionFlags{ WINHTTP_FLAG_ASYNC }
+  , m_sessionHandle{ NULL }
+  , m_connectionHandle{ NULL }
+  , m_requestHandle{ NULL }
+  , m_errorStatus{ ERROR_SUCCESS }
+{
+#if 0 // WinHttpCrackUrl does not accept non-HTTP schemes 🙄
+  BSTR bstr = _com_util::ConvertStringToBSTR(urlString.c_str());
+  LPWSTR lpwstr = bstr;
+  SysFreeString(bstr);
+
+  LPURL_COMPONENTS url{};
+  auto parseResult = WinHttpCrackUrl(lpwstr, static_cast<DWORD>(urlString.length()), /*flags*/ 0, url);
+  if (!parseResult)
+    throw std::exception("Could not parse URL.");
+
+  if (wcscmp(L"wss", url->lpszScheme) == 0)
+  {
+    m_openSessionFlags |= WINHTTP_FLAG_SECURE;
+  }
+  else if (wcscmp(L"ws", url->lpszScheme) != 0)
+  {
+    throw std::exception(Common::Unicode::Utf16ToUtf8(
+      std::wstring(L"Incorrect URL scheme: ") + std::wstring(url->lpszScheme)
+    ).c_str());
+  }
+
+  m_url = *url;
+#else
+  Url url{ urlString };
+  ZeroMemory(&m_url, sizeof(m_url));
```

Comment:
>ZeroMemory [](start = 2, length = 10)

why is this needed? shouldn't m_url constructor already take care of this?

Reply:
I'd think so, but this is based on MS Docs and official samples.

Reply:
there is no constructor, this is a Win32-style C type

---
In reply to: [393887322](https://github.com/microsoft/react-native-windows/pull/4330#discussion_r393887322) [](ancestors = 393887322)